File: backend/services/user-profiling-service/src/main/java/com/sudheer/robinhood/profiling/service/ProfilingService.java
package com.sudheer.robinhood.profiling.service;

import com.sudheer.robinhood.profiling.dto.QuickRiskAssessmentRequest;
import com.sudheer.robinhood.profiling.dto.QuickRiskAssessmentResponse;
import com.sudheer.robinhood.profiling.dto.StrategyRecommendation;
import com.sudheer.robinhood.profiling.enums.RiskTolerance;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class ProfilingService {

    public QuickRiskAssessmentResponse processQuickAssessment(QuickRiskAssessmentRequest request) {
        // Calculate risk score
        Integer riskScore = request.calculateRiskScore();

        // Map to risk tolerance
        RiskTolerance riskTolerance;
        String recommendedStrategy;

        if (riskScore < 30) {
            riskTolerance = RiskTolerance.CONSERVATIVE;
            recommendedStrategy = "Conservative Income";
        } else if (riskScore < 60) {
            riskTolerance = RiskTolerance.MODERATE;
            recommendedStrategy = "Balanced Growth";
        } else if (riskScore < 80) {
            riskTolerance = RiskTolerance.AGGRESSIVE;
            recommendedStrategy = "Growth Portfolio";
        } else {
            riskTolerance = RiskTolerance.VERY_AGGRESSIVE;
            recommendedStrategy = "Aggressive Growth";
        }

        log.info("Quick assessment processed: riskScore={}, tolerance={}", riskScore, riskTolerance);

        return QuickRiskAssessmentResponse.builder()
                .riskScore(riskScore)
                .riskTolerance(riskTolerance)
                .recommendedStrategy(recommendedStrategy)
                .assessment(request)
                .build();
    }

    public List<StrategyRecommendation> getRecommendations(Integer riskScore) {
        List<StrategyRecommendation> recommendations = new ArrayList<>();

        if (riskScore < 30) {
            // Conservative
            recommendations.add(StrategyRecommendation.builder()
                    .strategyName("Conservative Income")
                    .strategyType("conservative")
                    .expectedReturn(new BigDecimal("4.5"))
                    .volatility(new BigDecimal("5.0"))
                    .stockAllocation(new BigDecimal("20.0"))
                    .bondAllocation(new BigDecimal("70.0"))
                    .alternativeAllocation(new BigDecimal("5.0"))
                    .cashAllocation(new BigDecimal("5.0"))
                    .reasoning("Low risk tolerance with focus on capital preservation and steady income")
                    .riskMatchScore(95)
                    .build());
        } else if (riskScore < 60) {
            // Moderate
            recommendations.add(StrategyRecommendation.builder()
                    .strategyName("Balanced Growth")
                    .strategyType("moderate")
                    .expectedReturn(new BigDecimal("7.5"))
                    .volatility(new BigDecimal("10.0"))
                    .stockAllocation(new BigDecimal("60.0"))
                    .bondAllocation(new BigDecimal("35.0"))
                    .alternativeAllocation(new BigDecimal("5.0"))
                    .cashAllocation(new BigDecimal("0.0"))
                    .reasoning("Balanced approach with moderate risk for steady growth")
                    .riskMatchScore(90)
                    .build());
        } else {
            // Aggressive
            recommendations.add(StrategyRecommendation.builder()
                    .strategyName("Aggressive Growth")
                    .strategyType("aggressive")
                    .expectedReturn(new BigDecimal("12.0"))
                    .volatility(new BigDecimal("18.0"))
                    .stockAllocation(new BigDecimal("85.0"))
                    .bondAllocation(new BigDecimal("10.0"))
                    .alternativeAllocation(new BigDecimal("5.0"))
                    .cashAllocation(new BigDecimal("0.0"))
                    .reasoning("High growth potential with higher volatility for long-term investors")
                    .riskMatchScore(92)
                    .build());
        }

        return recommendations;
    }
}

File: frontend/src/services/profilingService.ts
import axios from 'axios';
import { USE_MOCK } from './config';

const API_BASE_URL = 'http://localhost:8082/api/v1/profiling';

interface QuickRiskAssessment {
    age: number;
    investmentAmount: number;
    investmentGoal: string;
    timeHorizonYears: number;
    lossTolerance: number;
    investmentKnowledge: number;
}

interface QuickAssessmentResponse {
    riskScore: number;
    riskTolerance: string;
    recommendedStrategy: string;
    assessment: QuickRiskAssessment;
}

interface StrategyRecommendation {
    strategyName: string;
    strategyType: string;
    expectedReturn: number;
    volatility: number;
    stockAllocation: number;
    bondAllocation: number;
    alternativeAllocation: number;
    cashAllocation: number;
    reasoning: string;
    riskMatchScore: number;
}

// Mock risk calculation based on user inputs
function calculateMockRiskScore(data: QuickRiskAssessment): number {
    const ageScore = Math.max(0, 100 - data.age);
    const lossScore = data.lossTolerance * 10;
    const knowledgeScore = data.investmentKnowledge * 10;
    const horizonScore = Math.min(data.timeHorizonYears * 3, 50);

    return Math.round((ageScore + lossScore + knowledgeScore + horizonScore) / 4);
}

function getRiskTolerance(score: number): string {
    if (score < 30) return 'CONSERVATIVE';
    if (score < 50) return 'MODERATE';
    if (score < 70) return 'AGGRESSIVE';
    return 'VERY_AGGRESSIVE';
}

function getRecommendedStrategy(tolerance: string): string {
    const strategies: Record<string, string> = {
        'CONSERVATIVE': 'Balanced Income',
        'MODERATE': 'Balanced Growth',
        'AGGRESSIVE': 'Growth Focus',
        'VERY_AGGRESSIVE': 'Aggressive Growth',
    };
    return strategies[tolerance] || 'Balanced Growth';
}

export const profilingService = {
    quickAssessment: async (data: QuickRiskAssessment): Promise<QuickAssessmentResponse> => {
        if (USE_MOCK) {
            await new Promise(resolve => setTimeout(resolve, 800));
            const riskScore = calculateMockRiskScore(data);
            const riskTolerance = getRiskTolerance(riskScore);
            return {
                riskScore,
                riskTolerance,
                recommendedStrategy: getRecommendedStrategy(riskTolerance),
                assessment: data,
            };
        }
        const response = await axios.post(`${API_BASE_URL}/quick-assessment`, data);
        return response.data;
    },

    getRecommendations: async (riskScore: number): Promise<StrategyRecommendation[]> => {
        if (USE_MOCK) {
            await new Promise(resolve => setTimeout(resolve, 500));
            return [
                {
                    strategyName: 'Balanced Growth',
                    strategyType: 'MODERATE',
                    expectedReturn: 8.5,
                    volatility: 12,
                    stockAllocation: 60,
                    bondAllocation: 30,
                    alternativeAllocation: 5,
                    cashAllocation: 5,
                    reasoning: 'Balanced approach for moderate risk tolerance',
                    riskMatchScore: riskScore,
                },
            ];
        }
        const response = await axios.get(`${API_BASE_URL}/recommendations?riskScore=${riskScore}`);
        return response.data;
    },

    getRiskToleranceOptions: async () => {
        if (USE_MOCK) {
            return ['CONSERVATIVE', 'MODERATE', 'AGGRESSIVE', 'VERY_AGGRESSIVE'];
        }
        const response = await axios.get(`${API_BASE_URL}/enums/risk-tolerance`);
        return response.data;
    },

    getInvestmentGoals: async () => {
        if (USE_MOCK) {
            return ['RETIREMENT', 'WEALTH_BUILDING', 'INCOME_GENERATION', 'CAPITAL_PRESERVATION', 'EDUCATION', 'HOME_PURCHASE'];
        }
        const response = await axios.get(`${API_BASE_URL}/enums/investment-goals`);
        return response.data;
    },
};

